[docs] create and send txs (#1361)
iotaledger/goshimmer: docs/tutorials/send_transaction.md

# How to send transaction
The simplest and easiest way for creating transaction is to use ready solution, such us GUI wallets: [pollen-wallet](https://github.com/iotaledger/pollen-wallet/tree/master) and [Dr-Electron ElectricShimmer](https://github.com/Dr-Electron/ElectricShimmer)
or command line wallet [Command Line Wallet](./wallet.md). However, there is also an option to create a transaction directly with Go client library, which will be main focus of this tutorial.

For the ready code examples you can go directly to [Code examples](send_transaction.md#ready-examples).

## Funds
To create a transaction, firstly we need to be in possession of tokens. We can receive them from other network users or request them from the faucet. For more details on how to request funds, see [this](./request_funds.md) tutorial.

## Preparing transaction

> A transaction is build from two main parts: a transaction essence, and an unlock block. The transaction essence contains all crucial information on how many and where the funds should be sent. The unlock block makes sure that only the owner of the destination address will be able to spend outputs of this transaction.

Reviewer: ```suggestion
A transaction is built from two parts: a transaction essence, and the unlock blocks. The transaction essence contains, among other information, the amount, the origin and where the funds should be sent. The unlock block makes sure that only the owner of the funds being transferred is allowed to successfully perform this transaction.
```

### Seed

> In order to send funds we need to have private key that can unlock our funds. If you want to use an existing seed from one of your wallets, just use the backup seed showed during a wallet creation. With this, we can decode the string with `base58` library and create the `seed.Seed` instance. That will allow us to retrieve wallet addresses (`mySeed.Address()`) and their corresponding private and public keys (`mySeed.KeyPair()`).

Reviewer: ```suggestion
In order to send funds we need to have a private key that can be used to prove that we own the funds and consequently unlock them. If you want to use an existing seed from one of your wallets, just use the backup seed showed during a wallet creation. With this, we can decode the string with the `base58` library and create the `seed.Seed` instance. That will allow us to retrieve the wallet addresses (`mySeed.Address()`) and the corresponding private and public keys (`mySeed.KeyPair()`).
```

```Go
balance, colorExist := out.Balances().Get(ledgerstate.ColorIOTA)
fmt.Println(balance, exist)
```
or iterate over all colors and balances:
```Go
out.Balances().ForEach(func(color ledgerstate.Color, balance uint64) bool {
			fmt.Println("Color:", color.Base58())
			fmt.Println("Balance:", balance)
			return true
		})
```

At the end we need to wrap the selected output to match the interface of the inputs:
```Go
inputs = ledgerstate.NewInputs(ledgerstate.NewUTXOInput(out))
```

#### Outputs
To create the most basic type of output use
`ledgerstate.NewSigLockedColoredOutput()` and provide it with a balance and destination address. Important is to provide the correct balance value. The total balance with the same color has to be equal for input and output.
```Go
balance := ledgerstate.NewColoredBalances(map[ledgerstate.Color]uint64{
							ledgerstate.ColorIOTA: uint64(100),
						})
outputs := ledgerstate.NewOutputs(ledgerstate.NewSigLockedColoredOutput(balance, destAddr.Address()))
```
The same as in case of inputs we need to adapt it with `ledgerstate.NewOutputs()` before passing to the `NewTransactionEssence` function.

### Signing transaction

> After preparing the transaction essence there is a time to sign it and put the signature to the unlock block part of the transaction.

Reviewer: ```suggestion
After preparing the transaction essence, we should sign it and put the signature to the unlock block part of the transaction.
```

We can retrieve private and public key pairs from the seed by providing it with indexes corresponding to the addresses that holds the unspent output that we want to use in our transaction.
```Go
keyPair := *mySeed.KeyPair(0)
txEssence := NewTransactionEssence(version, timestamp, accessPledgeID, consensusPledgeID, inputs, outputs)
```
We can sign the transaction in two ways: with ED25519 or BLS signature. The wallet seed library uses `ed25519` package and keys, so we will use `Sign()` method along with `ledgerstate.ED25519Signature` constructor to sign the transaction essence bytes.
Next step is to create the unlock block from our signature.

```Go
signature := ledgerstate.NewED25519Signature(kp.PublicKey, kp.PrivateKey.Sign(txEssence.Bytes())
unlockBlock := ledgerstate.NewSignatureUnlockBlock(signature)
```
Putting it all together, now we are able to create transaction with previously created transaction essence and adapted unlock block.

```Go
tx := ledgerstate.NewTransaction(txEssence, ledgerstate.UnlockBlocks{unlockBlock})
```

## Sending transaction
There are two web API methods that allows us to send the transaction:

> PostTransaction() and IssuePayload(). The second one is more general method that sends the attached payload. We are going to use the first one that will additionally checks the transaction validity before issuing and wait with sending the response until the message is booked.

Reviewer 2: ```suggestion
`PostTransaction()` and `IssuePayload()`. The second one is a more general method that sends the attached payload. We are going to use the first one that will additionally check the transaction validity before issuing and wait with sending the response until the message is booked.
```

	// sign.
	seed := walletseed.NewSeed([]byte("your_seed"))
	kp := seed.KeyPair(0)
	sig := ledgerstate.NewED25519Signature(kp.PublicKey, kp.PrivateKey.Sign(txEssence.Bytes()))
	unlockBlock := ledgerstate.NewSignatureUnlockBlock(sig)

	// build tx.
	tx = ledgerstate.NewTransaction(txEssence, ledgerstate.UnlockBlocks{unlockBlock})
	return
}
```

### Post the transaction

There are 2 available options to post the created transaction.

 - Goshimmer client lib
 - Web API
 
#### Post via client lib
```go
func postTransactionViaClientLib() (res string , err error) {
	// connect to goshimmer node
	goshimmerClient := client.NewGoShimmerAPI("http://127.0.0.1:8080", client.WithHTTPClient(http.Client{Timeout: 60 * time.Second}))

	// build tx from previous step
	tx, err := buildTransaction()
	if err != nil {
		return
	}

	// send the tx payload.
	res, err = goshimmerClient.PostTransaction(tx.Bytes())
	if err != nil {
		return
	}
	return
}
```

#### Post via web API
First, get the transaction bytes.
```go
	// build tx from previous step
	tx, err := buildTransaction()
	if err != nil {
		return

> 	}

Reviewer 2: Adjust indent.
